crates/shim/src/lib.rs

```diff
@@ -84,45 +87,52 @@ pub struct StartOpts {
 }
 
 /// Helper structure that wraps atomic bool to signal shim server when to shutdown the TTRPC server.
+///
 /// Shim implementations are responsible for calling [`Self::signal`].
 #[derive(Clone)]
-pub struct ExitSignal(Arc<AtomicBool>);
+pub struct ExitSignal(Arc<(Mutex<bool>, Condvar)>);
 
 impl Default for ExitSignal {
+    #[allow(clippy::mutex_atomic)]
     fn default() -> Self {
-        ExitSignal(Arc::new(AtomicBool::new(false)))
+        ExitSignal(Arc::new((Mutex::new(false), Condvar::new())))
     }
 }
 
 impl ExitSignal {
     /// Set exit signal to shutdown shim server.
     pub fn signal(&self) {
-        self.0.store(true, Ordering::Release)
+        let (lock, cvar) = &*self.0;
+        let mut exit = lock.lock().unwrap();
+        *exit = true;
+        cvar.notify_all();
     }
 
     /// Wait for the exit signal to be set.
-    fn wait(&self) {
-        while !self.0.load(Ordering::Acquire) {
-            std::hint::spin_loop();
+    pub fn wait(&self) {
+        let (lock, cvar) = &*self.0;
+        let mut started = lock.lock().unwrap();
+        while !*started {
+            started = cvar.wait(started).unwrap();
         }
     }
 }
 
 /// Main shim interface that must be implemented by all shims.
+///
 /// Start and delete routines will be called to handle containerd's shim lifecycle requests.
-pub trait Shim: Task {
+pub trait Shim {
     /// Error type to be returned when starting/deleting shim.
     type Error: error::Error;
 
-    fn new(
-        id: &str,
-        namespace: &str,
-        publisher: RemotePublisher,
-        config: &mut Config,
-        exit: ExitSignal,
-    ) -> Self;
+    /// Type to provide task service for the shim.
+    type T: Task + Send + Sync;
```

Comment:
I think we should move away from the concept that there 1:1 relation between shim and task service.
We started adding plugin support to shims, so any number of arbitrary service can now be registered (https://github.com/containerd/containerd/pull/5817, https://github.com/containerd/containerd/pull/6021).

Reply:
Exactly, the same reason to refactor Shim interfaces.